ext/nnstreamer/tensor_filter/vivante/tensor_filter_subplugin.c

```diff
@@ -0,0 +1,594 @@
+vivante_close (const GstTensorFilterProperties * prop,
+    void **private_data);
+
+/**
+ *  * @brief Configure private_data
+ *   */
+static int
+allocateData (const GstTensorFilterProperties * prop, void **private_data)
+{
+  vivante_pdata *pdata;
+
+  if (*private_data != NULL) {
+    /* Already opened */
+    pdata = (vivante_pdata *) *private_data;
+
+    if (!prop->model_files[0] || prop->model_files[0][0] == '\0') {
+      printf("Model path (.nb) is not given.");
+      return -1;
+    }
+    if (!prop->model_files[1] || prop->model_files[1][0] == '\0') {
+      printf("Shared library path (.so) is not given.");
+      return -1;
+    }
+    if (pdata->model_path && g_strcmp0 (prop->model_files[0],
+          pdata->model_path) == 0) {
+      return 0; /* Already opened with same model file. Skip ops */
+    }
+    if (pdata->so_path && g_strcmp0 (prop->model_files[1],
+          pdata->so_path) == 0) {
+      return 0; /* Already opened with same so file. Skip ops */
+    }
+    vivante_close (prop, private_data); /* Close before opening one. */
+  }
+
+  *private_data = pdata = g_new0 (vivante_pdata, 1);
+  if (pdata == NULL) {
+      printf("Failed to allocate memory for vivante tensor_filer.");
+    return -1;
+  }
+
+  return 0;
+}
+
+/**
+ * @brief Convert a tesor type from vivant format to nnstreamer format
+ * Note that you must refer to the "linux_sdk/acuity-ovxlib-dev/include/vsi_nn_types.h" file.
+ */
+unsigned int
+convert_tensortype (unsigned tensor_type)
+{
+  switch (tensor_type) {
+  case VSI_NN_TYPE_INT8:
+    return _NNS_INT8;
+  case VSI_NN_TYPE_UINT8:
+    return _NNS_UINT8;
+  case VSI_NN_TYPE_INT16:
+    return _NNS_INT16;
+  case VSI_NN_TYPE_UINT16:
+    return _NNS_UINT16;
+  /** Note that the current nnstreamer (tensor_typedef.h) does not support FLOAT16.
+   *  Let's use UINT16 as a workaround.
+   */
+  case VSI_NN_TYPE_FLOAT16:
+    return _NNS_UINT16;
+  case VSI_NN_TYPE_FLOAT32:
+    return _NNS_FLOAT32;
+  default:
+    /** @todo Support other types */
+    break;
+  }
+  return _NNS_END;
+}
+
+/**
+ * @brief Open the vivante tensor filter
+ * @note
+ * 1. Read multi files (e.g., inception_v3.nb and libinception_v3.so)
+ */
+static int
+vivante_open (const GstTensorFilterProperties * prop, void **private_data)
+{
+  int ret = allocateData (prop, private_data);
+  unsigned int i,j,k;
+  vivante_pdata *pdata = (vivante_pdata *) *private_data;
+
+  if (ret < 0)
+    return ret;
+
+  if (*private_data == NULL)
+    return -ENOMEM;
+
+  pdata->model_path = g_strdup (prop->model_files[0]);
+  pdata->so_path    = g_strdup (prop->model_files[1]);
+
+  ret = parseCustomProperty (prop->custom_properties, &pdata->postProcess);
+  if (ret < 0)
+    return ret;
+
+  /** Create the neural network with .nb (a network binary of Vivante) */
+  pdata->handle = dlopen(pdata->so_path, RTLD_NOW);
+  if (!pdata->handle) {
+    printf ("vivante_open: dlopen cannot load the shared library (.so).\n");
+    return -EINVAL;
```

Comment:
Possible memory leaks due to line 277~278

Reply:
Do you mean g_free()?

Reply:
Yes, you may need `g_free()` for duplicated strings

Reply:
Thanks. Fixed. :)

Reply:
Still missing `g_free ()` in line 280.. T_T

Reply:
@leemgs PTAL here.

Reply:
Ooops. Fixed. :)